src/next/scval_conversions.rs

```diff
@@ -166,36 +168,51 @@ pub mod int128_helpers {
     #[must_use]
     #[inline(always)]
     #[allow(clippy::inline_always, clippy::cast_possible_truncation)]
-    pub fn u128_lo(u: u128) -> u64 {
-        u as u64
+    pub fn u128_hi(u: u128) -> u64 {
+        (u >> 64) as u64
     }
 
     #[must_use]
     #[inline(always)]
     #[allow(clippy::inline_always, clippy::cast_possible_truncation)]
-    pub fn u128_hi(u: u128) -> u64 {
-        (u >> 64) as u64
+    pub fn u128_lo(u: u128) -> u64 {
+        u as u64
     }
 
     #[must_use]
     #[inline(always)]
     #[allow(clippy::inline_always)]
-    pub fn u128_from_pieces(lo: u64, hi: u64) -> u128 {
-        u128::from(lo) | (u128::from(hi) << 64)
+    pub fn u128_from_pieces(hi: u64, lo: u64) -> u128 {
+        (u128::from(hi) << 64) | u128::from(lo)
+    }
+
+    #[must_use]
+    #[inline(always)]
+    #[allow(clippy::inline_always, clippy::cast_possible_truncation)]
+    pub fn i128_hi(i: i128) -> i64 {
+        (i >> 64) as i64
     }
 
     #[must_use]
     #[inline(always)]
-    #[allow(clippy::inline_always, clippy::cast_sign_loss)]
-    pub fn u128_from_i128(i: i128) -> u128 {
-        i as u128
+    #[allow(
+        clippy::inline_always,
+        clippy::cast_possible_truncation,
+        clippy::cast_sign_loss
+    )]
+    pub fn i128_lo(i: i128) -> u64 {
+        i as u64
     }
 
     #[must_use]
     #[inline(always)]
-    #[allow(clippy::inline_always, clippy::cast_possible_wrap)]
-    pub fn i128_from_u128(u: u128) -> i128 {
-        u as i128
+    #[allow(
+        clippy::inline_always,
+        clippy::cast_sign_loss,
+        clippy::cast_possible_wrap
+    )]
+    pub fn i128_from_pieces(hi: i64, lo: u64) -> i128 {
+        ((hi as u128) << 64 | u128::from(lo)) as i128
```

Comment:
I don't actually know what happens when you do a direct i64-to-u128 conversion in Rust as you're doing in the first expression `(hi as u128)`. I _think_ it does a sign extending i64-to-i128, followed by a sign-change "no-op" i128-to-u128 that does not change any bits. But I'm not certain. It might also do a sign-change no-op from i64-to-u64 followed by a zero-extension u64-to-u128. Which, if that's what it does, will be wrong here. The [official semantics are here](https://doc.rust-lang.org/reference/expressions/operator-expr.html#semantics) and suggest your code is doing the right thing, but I'm still not 100% certain.

My rule of thumb for making sure I don't make such errors in Rust is to do only one such action at a time: either switch signedness without changing width, or change width without changing signedness. And if you have to do both, take an extra nesting level of expressions or line of code (it all compiles away anyways) and do the two steps in two separate expressions or statements.

Reply:
Good point, I will take the extra step to make those conversions more explicit. 
(I've referred to the official doc as well, but I agree it's good to be extra-safe)

Reply:
I have changed this line, hope it's more clear now.

Reply:
(FWIW, I found the [official rust semantics](https://doc.rust-lang.org/reference/expressions/operator-expr.html#numeric-cast:~:text=Casting%20from%20a%20smaller,the%20source%20is%20signed) quite easy to follow. Casting from a smaller to a larger int, it just sign extends the source (smaller) int.)